### PR TITLE
fix(native): carry a long double's full significand into fixpnt, areal, rational and lns

### LINKED ISSUE #1517: native: carry a binary128 long double's full 113-bit significand into wide and exact types
## Summary

#1515 / #1516 fixed long double assignment on binary128 (aarch64 Linux) and IBM double-double (POWER). `extractFields()` now hands out x87-shaped fields computed from the value: the significand is cut to **64 bits**, and the rest is folded into a round-to-odd sticky bit. That is correct for every target of 62 significand bits or fewer. A wider or exact target gets the long double's leading 64 bits instead of its value:

- `cfloat<128,15>` (the binary128 layout) = a 113-bit long double loses bits 65-113.
- `einteger` / `edecimal` / `erational` / `rational` = 2^100 + 1 (exact in binary128) does not come out exact.
- `posit<128,...>`, `fixpnt` with more than 62 significant bits, `lns` with many rbits, and so on.

On x87 (64-bit significand) nothing is lost, and on double-double the leading 64 of ~106 bits are kept.

## Why it was left out of #1516

Every caller carries the fraction in a single `uint64_t` and shifts by `ieee754_parameter<long double>::fbits`. Carrying 112 bits means a new extraction API (two words, or `integer<>` / `blockbinary`) and rounding code in each type's `convert_ieee754()`: about 10 types.

## Suggested direction

- An extraction that returns the full significand, e.g. `{sign, class, unbiased exponent, blockbinary<128> significand, sticky}`, computed from the value like `extractLongDoubleFields()` so it covers binary128 and double-double alike. (Double-double values can span more than 106 bits with a gap, so take two 64-bit cuts plus sticky.)
- Types that can hold more than 64 bits use it: cfloat when `fbits > 62`, the exact/adaptive types (einteger, edecimal, erational, rational), posit/fixpnt/lns above 62 bits.
- Keep the 64-bit round-to-odd path for narrow targets.

## Suggested test

Extend `static/conversions/long_double_conversion.cpp`:
- round-trip every long double through `cfloat<128,15>` (the encoding can be checked against a `frexp`-built one, as in `static/float/cfloat/conversion/native_conversion.cpp`);
- `einteger`, `rational` and `erational` of 2^100 + 1 and of 1 + 2^-100 must be exact;
- run under `qemu-aarch64-static` and `qemu-ppc64le-static`, and on native Linux ARM64 CI.

Related: #1515, #1516, #1514.

### LINKED ISSUE #1519: rational: assignment from a native float saturates past scale 64 and prints 'overflow' to stdout
## Summary

On every platform, `rational<128>` assigned 2^70 comes out as maxpos (2^127), and `overflow: scale = 70` is printed to stdout:

```cpp
rational<128> r;
r = std::ldexp(1.0, 70);   // prints "overflow: scale = 70"; r == 1.70141e+38 (2^127), want 1.18059e+21
```

Measured on `main` (9e085028), x86_64, gcc 13.

## Cause

`rational::convert_ieee754()` (`include/sw/universal/number/rational/rational_impl.hpp`, around lines 382-480) works entirely in `uint64_t`:
- `a = f | hmask`, `b = hmask`;
- a gcd on uint64;
- `if (scale >= 64)` treated as overflow or underflow, with a diagnostic print;
- `n = static_cast<int64_t>(a)`.

So any value whose scale needs more than 64 bits saturates, even when `rational<nbits>` could hold it. The print is library output on a normal code path.

## Suggested fix

Do the conversion in the rational's own `blockbinary<nbits>` width:
- the numerator is the significand, and the denominator a power of two;
- reduce by counting trailing zeros (the gcd with a power of two);
- shift within nbits, saturating only when nbits really overflows;
- drop the stdout print.

#1517's second PR rewrites this function to carry a long double's full significand, and can close this issue.

## Suggested test

`rational<128>` of 2^k for k in [60, 120] and of 2^-k in [60, 120], plus odd significands (3 * 2^70), from double and long double. Expect exact values and no output.

Found by the #1517 audit.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary
Part 2 of #1517; part 1 was #1520. fixpnt, areal, rational and lns still took the leading 64 bits `extractFields()` gives, whose last is a round-to-odd sticky bit on IEEE binary128 and IBM double-double. They now read the whole significand through `long_double_significand`.

| type | before (aarch64) | after |
|---|---|---|
| fixpnt<128,64> = 2^40 + 2^-60 | lost the 2^-60, and ties mis-rounded | exact, RNE ties |
| areal<128,15> = 1 + 2^-100 | the interval (1, 1 + 2^-63) | exact (ubit clear) |
| areal<80,15> = 1 + 2^-100 | | 1 with the ubit set: it has 64 fraction bits |
| rational<128> = 1 + 2^-100 | not exact | exact |
| lns, more than 63 bits | log2 rounded from 64 bits | from every bit |

## Rational, on every platform, from double as well
Rational's conversion ran in a `uint64_t`. It now runs **the same algorithm** on an integer wide enough for nbits and a long double's significand:
1. reduce by the common power of two;
2. shift to nbits;
3. shift up by the scale.

Small rationals therefore convert exactly as before; the exhaustive `rational_assignment` test, which requires `r = double(r)` to give `r` back, still passes. The widened version also fixes these, from plain doubles:
- **#1519**: `rational<128> = 2^70` became maxpos, and printed `overflow: scale = 70` to stdout. The prints are gone.
- `rational<8> = -1.0` gave **+1**.
- a negative overflow saturated to **maxpos** instead of maxneg.
- a subnormal double was left unconverted: `rational<16> = 2^-1030` gave 1.11641.

## Changes
- `fixpnt_impl.hpp`: `convert_long_double()` for nbits > 62. Bit j lands at position `scale + rbits - j`. Positions from nbits up wrap away as for a double; below 0 come the guard and sticky bits. RNE on the magnitude, then negate.
- `areal_impl.hpp`: `assign_extended()` reads a word at a time until the value is read or the blocktriple is full. Only bits that don't fit set the ubit. It stays constexpr and frexp-free.
- `rational_impl.hpp`: the widened conversion core.
- `lns_impl.hpp`: `log_exponent_from_long_double()` for nbits − 1 > 62 (lns requires rbits < 64).
- `static/conversions/long_double_conversion.cpp` adds:
  - fixpnt<128,64> exact and at ties;
  - areal<128,15> exact, and areal<80,15> with the ubit;
  - rational<128> exact and past scale 64;
  - the rational sign and subnormal cases;
  - lns<80,60> exponent bits against `rint(log2l(v) * 2^60)`.

## Test Results
The new checks against part 1's headers: **19 failures on x86** (the rational fixes) and **34 under qemu-aarch64**.

| config | long double | long_double_conversion (11 groups) |
|---|---|---|
| x86 gcc, x86 clang | x87 | PASS |
| aarch64 gcc, aarch64 clang (qemu) | binary128 | PASS |
| ppc64le gcc (qemu) | double-double | PASS |

Suites for cfloat, posit, lns, dbns, fixpnt, rational, takum, areal, einteger, edecimal, erational, unum, internals, native and conversions: **400/400 on gcc and on clang**, no new warnings. `check-ascii.sh` passes.

Resolves #1517
Resolves #1519

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved `long double` conversions for wide fixed-point, rational, logarithmic, and adaptive real-number formats.
  * Preserved significand bits beyond the 64-bit boundary to improve conversion precision.
  * Corrected rounding, truncation, sign handling, overflow, underflow, subnormal conversion, and uncertainty reporting.

* **Tests**
  * Added regression coverage for wide values, low-order bits, boundary conditions, and signed conversions.

### CODE REVIEW

File: include/sw/universal/number/rational/rational_impl.hpp [hunk 1]
```diff
@@ -388,98 +388,125 @@ class rational<_nbits, base2, bt> {
...
+
+		if (exponent == 0 && a == b) {
+			n = (s ? -1 : 1);
+			d = 1;
+			return *this;
+		}
+		if (exponent >= 0) {
+			unsigned msb = msb1(a);
+			if (msb > nbits) {
+				const unsigned shift = 1u + msb - nbits;
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/rational/rational_impl.hpp]: Confirmed and fixed in 0038632b. Both boundaries were kept unchanged from the uint64_t version and are on main too: rational<8> = 255.0 gave 1, and 128.0 and 200.0 gave NaN with "rational_divide_by_zero". Both width guards and both scale guards are now >=. Over every rational<8> value converted through its double, exactly two results change, 128 and -128: they were NaN and are now maxpos and -128/1. long_double_conversion.cpp now checks 255, 128, 200 and 1e10 against maxpos, and -128 against maxneg. Verified on x86 with gcc and clang, and under qemu-aarch64 and qemu-ppc64le.

File: include/sw/universal/number/rational/rational_impl.hpp [hunk 2]
```diff
@@ -388,98 +388,125 @@ class rational<_nbits, base2, bt> {
...
+				b >>= static_cast<int>(shift);
+			}
+			msb = msb1(a);
+			const uint64_t maxUpShift   = (nbits - msb - 1u);
+			const uint64_t maxDownShift = msb1(b);
+			const uint64_t scale        = static_cast<uint64_t>(exponent);
+			if (scale > maxUpShift) {
+				if (scale > (maxUpShift + maxDownShift)) {  // too large for nbits
+					if (s) maxneg(); else maxpos();
+					return *this;
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/rational/rational_impl.hpp]: Confirmed and fixed in 0038632b. Both boundaries were kept unchanged from the uint64_t version and are on main too: rational<8> = 255.0 gave 1, and 128.0 and 200.0 gave NaN with "rational_divide_by_zero". Both width guards and both scale guards are now >=. Over every rational<8> value converted through its double, exactly two results change, 128 and -128: they were NaN and are now maxpos and -128/1. long_double_conversion.cpp now checks 255, 128, 200 and 1e10 against maxpos, and -128 against maxneg. Verified on x86 with gcc and clang, and under qemu-aarch64 and qemu-ppc64le.
